woocommerce/woocommerce-android
# Fix/9654 local notification fix

Closes: #9654


### Description

Fixes several issues related to local notifications: 
- Avoid canceling scheduled local notifications when the user switches to a different store.
- Switch to the correct store upon tapping on a local notification the belongs to another site.
- ~~Notifications missing the `siteId` were being scheduled but never displayed due to the logic in `PreconditionCheckWorker`:~~ 
```kotlin
    private fun proceedIfFreeTrialAndMatchesSite(siteId: Long?): Result {
        val site = selectedSite.get()
        return if (site.isFreeTrial && site.siteId == siteId) {
            Result.success()
        } else {
            cancelWork("Store plan upgraded or a different site. Cancelling work.")
        }
    }
```
~~This PR fixes that and ensures all the scheduled local notifications have the correct `siteId` so we can support site switching.~~
EDIT: As discussed below https://github.com/woocommerce/woocommerce-android/pull/9736#issuecomment-1709798125 not all local notifications are impacted by the previous code. And the ones that were impacted by the previous code had the proper `siteId` set. 

### Testing instructions

The changes in this PR mainly affect `StoreCreationCompletedNotification` so focusing our test around that notification should be enough.  
1. First apply the following change to make the notification trigger sooner after being scheduled: 
``` diff
Subject: [PATCH] Wait only 2 min for notification
---
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/notifications/local/LocalNotification.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/notifications/local/LocalNotification.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/notifications/local/LocalNotification.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/notifications/local/LocalNotification.kt	(revision a451a5453f1e7d721de5c9953805585f548cde03)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/notifications/local/LocalNotification.kt	(date 1694072038449)
@@ -30,7 +30,7 @@
         title = R.string.local_notification_store_creation_complete_title,
         description = R.string.local_notification_store_creation_complete_description,
         type = LocalNotificationType.STORE_CREATION_FINISHED,
-        delay = 5,
+        delay = 2,
         delayUnit = TimeUnit.MINUTES
     ) {
 
```
2. Trigger store creation flow and press "Try for free". Wait for the loading to complete until you are taken to the first profiler step. You should see the following log before moving to the next step: 
```
Tracked: local_notification_scheduled, Properties: {"type":"store_creation_complete","blog_id":223209821,"is_wpcom_store":true,"was_ecommerce_trial":true,"plan_product_slug":"ecommerce-trial-bundle-monthly","is_debug":true,"site_url":"url"}
```
3. Kill the app
4. Wait for the notification to show up. We reduced the waiting time to 2 mins, it should be enough for the store to be ready.
5. Tap on the notification and notice how the stores is switched to the new one created one. 
6. Review the following event is only tracked once: 
```
Tracked: local_notification_tapped, Properties: {"type":"store_creation_complete","blog_id":223232803,"is_wpcom_store":true,"was_ecommerce_trial":true,"plan_product_slug":"ecommerce-trial-bundle-monthly","is_debug":true,"site_url":"url"}
```
